File: nbs_compression/sia_family/sia.py
from .point import Point, Vector

from typing import List, Dict


def sia(dataset: List[Point], restrict_dpitch_zero: bool = False) -> Dict[Vector, List[Point]]:
    """
    SIA algorithm: compute all maximal translatable patterns (MTPs).
    Returns a dict: vector -> list of points that are the MTP (starting points).
    The MTP for vector v is { p in dataset | p+v in dataset }.
    """
    # sort dataset lexicographically (by tick, then pitch)
    points = sorted(dataset)
    n = len(points)
    # build vector table: for each i>j, vector = points[i] - points[j]
    # we'll store (vector, start_point) pairs
    vectors = []  # list of (v, p_start)
    for i in range(n):
        for j in range(n):
            if i == j:
                continue
            if points[i].instrument != points[j].instrument: 
                continue
            v = (points[i].tick - points[j].tick, points[i].key - points[j].key)
            # filter non-zero dpitch
            if restrict_dpitch_zero and v[1] != 0:
                continue
            vectors.append((v, points[j]))
    # sort by vector (lexicographically)
    vectors.sort(key=lambda x: x[0])
    # scan to group by vector
    mtps = {}
    i = 0
    while i < len(vectors):
        v = vectors[i][0]
        start_points = []
        while i < len(vectors) and vectors[i][0] == v:
            start_points.append(vectors[i][1])
            i += 1
        # remove duplicates (same start point may appear multiple times from different pairs)
        start_points = sorted(set(start_points))
        
        if v == (0, 0):
            continue
        if len(start_points) < 2:
            continue
        
        mtps[v] = start_points
    return mtps

File: src/sia.rs
use std::collections::HashMap;
use std::collections::HashSet;


/// Finds all maximal translatable patterns (MTPs) in a 2‑D point set using the SIA algorithm.
///
/// SIA (Structure Induction Algorithm) computes, for every non‑zero translation vector `v`,
/// the set of start points `p` such that both `p` and `p + v` belong to the dataset.
/// This set is the maximal translatable pattern for `v`. The algorithm groups all
/// ordered point pairs `(p, q)` by the vector `q - p`, then collects the starting points
/// for each vector.
///
/// # Arguments
/// * `dataset` - A reference to a vector of `(tick, pitch)` points with non‑negative coordinates.
/// * `restrict_dpitch_zero` - If `true`, only vectors with zero pitch difference are kept
///   (i.e., purely temporal translations). This restricts patterns to horizontal repetition.
///
/// # Returns
/// A vector of `(vector, pattern)` entries, where:
/// - `vector` is `(dtick, dpitch)` – the translation vector.
/// - `pattern` is a sorted list of points that form the MTP for that vector.
///
/// The zero vector `(0, 0)` is always excluded. Vectors that map fewer than two points
/// (i.e., pattern size < 2) are also omitted because a pattern must have at least one
/// translation to be translatable.
///
/// # Complexity
/// - Time: `O(n²)` in the worst case, where `n` is the number of points.
/// - Memory: `O(n²)` in the worst case (stores a set of start points for each distinct vector).
pub fn find_mtps(
    dataset: &Vec<(u32, u32)>, 
    restrict_dpitch_zero: bool
) -> HashMap<(i32, i32), Vec<(u32, u32)>> {
    let points = dataset;
    let n = points.len();

    // Online grouping using HashMap to avoid storing all O(n²) pairs
    let mut groups: HashMap<(i32, i32), HashSet<(u32, u32)>> = HashMap::new();

    for i in 0..n {
        let (ti, pi) = points[i];
        for j in 0..i {            
            let (tj, pj) = points[j];
            // Compute vector (i32 range, differences may be negative)
            let dx = ti as i32 - tj as i32;
            let dy = pi as i32 - pj as i32;
            if restrict_dpitch_zero && dy != 0 {
                continue;
            }
            // Insert into group with pj as the starting point
            groups
                .entry((dx, dy))
                .or_insert_with(HashSet::new)
                .insert((tj, pj));
            groups
                .entry((-dx, -dy))
                .or_insert_with(HashSet::new)
                .insert((ti, pi));
        }
    }

    // Collect results, filtering out zero vector and groups with fewer than 2 start points
    let mut result = HashMap::new();
    for (vec, start_set) in groups {
        if vec == (0, 0) {
            continue;
        }
        if start_set.len() < 2 {
            continue;
        }
        // Convert to sorted Vec
        let mut start_points: Vec<_> = start_set.into_iter().collect();
        start_points.sort();                 // sort by (tick, pitch)
        result.insert(vec, start_points);
    }
    result
}